sdk/src/types/block/slot/epoch.rs

```diff
@@ -74,6 +74,20 @@ impl EpochIndex {
     pub fn last_slot_index(self, slots_per_epoch_exponent: u8) -> SlotIndex {
         SlotIndex::from_epoch_index(self + 1, slots_per_epoch_exponent) - 1
     }
+
+    /// Returns the slot for which to be eligible for the committee selection of epoch n, a potential validator must
+    /// issue at least one block between Registration Slot(n)+1. and this slot.
+    pub fn activity_window_slot(&self, slots_per_epoch_exponent: u8, epoch_nearing_threshold: u32) -> SlotIndex {
+        self.first_slot_index(slots_per_epoch_exponent) - epoch_nearing_threshold - 1
+    }
+
+    /// Returns the slot at which delegation can happen for the following epoch.
+    pub fn registration_slot(&self, slots_per_epoch_exponent: u8, epoch_nearing_threshold: u32) -> SlotIndex {
+        self.first_slot_index(slots_per_epoch_exponent)
```

Comment:
so what if this underflows?

Reply:
Good question, asked here https://github.com/iotaledger/tips/pull/141/files#r1378094011

Reply:
https://github.com/iotaledger/tips/pull/141#discussion_r1426565752

Reply:
Looks like we need syntactic checks for protocol params now lol

Reply:
Was this addressed?

Reply:
That's the lastest state https://github.com/iotaledger/tips/pull/141#discussion_r1467425812, shouldn't block this PR

Reply:
Ok we can update if necessary when he answers. Please follow-up with him.